tmt/utils.py

```diff
@@ -1006,8 +1006,93 @@ def environment_to_dict(variables: Union[str, List[str]]) -> EnvironmentType:
     return result
 
 
-def environment_file_to_dict(
-        env_files: Iterable[str], root: str = ".") -> Dict[str, str]:
+@lru_cache(maxsize=None)
+def environment_file_to_dict(env_file: str, root: str = ".") -> EnvironmentType:
+    """
+    Create dict from a file.
+
+    File should be in yaml/yml or dotenv format.
+
+    dotenv file example:
+        ```bash
+        A=B
+        C=D
+        ```
+
+    yaml file example:
+        ```yaml
+        A: B
+        C: D
+        ```
+
+    Path to the file should be relative to the metadata tree root.
+    """
+
+    env_file = env_file.strip()
+
+    # Fetch a remote file
+    if env_file.startswith("http"):
+        # Create retry session for longer retries, see #1229
+        session = retry_session.create(
+            retries=ENVFILE_RETRY_SESSION_RETRIES,
+            backoff_factor=ENVFILE_RETRY_SESSION_BACKOFF_FACTOR,
+            allowed_methods=('GET',),
+            status_forcelist=(
+                429,  # Too Many Requests
+                500,  # Internal Server Error
+                502,  # Bad Gateway
+                503,  # Service Unavailable
+                504   # Gateway Timeout
+                ),
+            )
+        try:
+            response = session.get(env_file)
+            response.raise_for_status()
+            content = response.text
+        except requests.RequestException as error:
+            raise GeneralError(
+                f"Failed to fetch the environment file from '{env_file}'. "
+                f"The problem was: '{error}'")
+
+    # Read a local file
+    else:
+        # Ensure we don't escape from the metadata tree root
+        try:
+            root_path = Path(root).resolve()
+            full_path = (Path(root_path) / Path(env_file)).resolve()
+            full_path.relative_to(root_path)
+        except ValueError:
+            raise GeneralError(
+                f"The 'environment-file' path '{full_path}' is outside "
+                f"of the metadata tree root '{root}'.")
+        if not Path(full_path).is_file():
+            raise GeneralError(f"File '{full_path}' doesn't exist.")
+
+        content = Path(full_path).read_text()
+
+    # Parse yaml file
+    if os.path.splitext(env_file)[1].lower() in ('.yaml', '.yml'):
+        environment = parse_yaml(content)
+
+    else:
+        try:
+            environment = parse_dotenv(content)
+
+        except ValueError:
+            raise GeneralError(
+                f"Failed to extract variables from environment file "
+                f"'{full_path}'. Ensure it has the proper format "
+                f"(i.e. A=B).")
+
+    if not environment:
+        log.warn(f"Empty environment file '{env_file}'.")
+
+        return {}
+
+    return environment
+
+
+def environment_files_to_dict(env_files: Iterable[str], root: str = ".") -> Dict[str, str]:
```

Comment:
IMO Need better name, this will be nightmare to type (_files_ vs _file_)

Why is it type `Dict[str, str]`  instead of `EnvironmentType`?

Reply:
WRT `Dict`: I think it's the original value, I fixed it in the single-file clone, forgot to fix the original too. Fixed.

WRT naming: I'm not sure there's a perfect solution, with "do for one" and "do for many" variants, the names will be very similar. And I'd like to preserve the "do for one" version, I like composability, it's easier to test both parts of the process, mocking, etc.

Reply:
Yeah, for testing it is much better. If only the difference can be more then one char in the middle of the name.